### ISSUE TITLE
vega-typings enum discussion

### ISSUE THREAD

**Issue by jonathanzong** [vega/vega#1562]
Hi everyone, I'd like to start a discussion about the possibility of using `enum`s in vega-typings.

There are quite a few types that are union types of a bunch of string constants. Examples:
https://github.com/vega/vega/blob/master/packages/vega-typings/types/spec/marktype.d.ts
https://github.com/vega/vega/blob/8941b87cbfd678618d76b6e0a99aa5f77c43abc3/packages/vega-typings/types/spec/scale.d.ts#L4

There are also some that almost conform to this pattern, but have additional members like `SignalRef`s.

For example: https://github.com/vega/vega/blob/8941b87cbfd678618d76b6e0a99aa5f77c43abc3/packages/vega-typings/types/spec/scale.d.ts#L29

This `SortOrder` example makes me think that we should switch to `enum` types for this use case, because in working on Lyra I noticed that if we adopt `vega-typings`, we currently have to either 1) use raw string literals or 2) keep a separate struct in Lyra to store the literal values. These will both easily go out of sync with `vega-typings` (they already have, because the `SortOrder` constants changed from `asc` to `ascending` and `desc` to `descending` between Vega 2 and today).

The benefit of using `enum`: they provide the type and the values at once. So, you could define an `enum SortOrderValues` and use `SortOrderValues.Asc` without worrying about the underlying constant, and you can also type variables as `x: SortOrderValues` to get the behavior we already have with union types.

tl;dr what if we e.g. replaced `SortOrder` with something like
```
export enum SortOrderValues {
  Asc = 'ascending',
  Desc = 'descending'
}

export type SortOrder = SortOrderValues | SignalRef;
```

**Comment by arvind** [comment]
cc @domoritz, @kanitw

**Comment by domoritz** [comment]
I think using raw string literals is actually a viable solution since the strong constants don't change usually. Moreover, the pattern you propose does not allow string constants anymore, which is a deal breaker for most users.

```
export enum SortOrderValues {
  Asc = 'ascending',
  Desc = 'descending'
}

export type SortOrder = SortOrderValues;

const a: SortOrderValues = 'ascending';
```

Vega-Lite uses this pattern:

```
export namespace SortOrderValues {
  export const Asc: 'ascending' = 'ascending';
  export const Desc: 'descending' = 'descending';
}

export type SortOrderValues = typeof SortOrderValues.Asc | typeof SortOrderValues.Desc;
```

But that's a lot more code so I'm not sure whether we want to go down that route for Vega.

**Comment by jonathanzong** [comment]
I see, that is a big problem with enums. I also found this pattern while digging around for info:

```
import { enumerate } from 'ts-transformer-enumerate';
import { SortOrder } from 'vega-typings';

type Filter<T, U> = T extends U ? T : never;
const SortOrderValues = enumerate<Filter<SortOrder, string>>();
```
```
const x: SortOrder = 'ascending';
const y: SortOrder = SortOrderValues.ascending;
```

in this case, `SortOrderValues` looks like
```
{
    ascending: "ascending";
    descending: "descending";
}
```

We could just do that in Lyra, and/or add it to typings if you think it's useful to have the consts in vega-typings

**Comment by domoritz** [comment]
That sounds like a good plan. Let's use it in Lyra first and maybe move over to Vega later if we think it provides enough value for more people.

**Comment by kanitw** [comment]
1) The transform that @jonathanzong  looks interesting.  Too bad that it still has this clause: 

> Unfortunately, TypeScript itself does not currently provide any easy way to use custom transformers

which means it's a bit harder to setup.   But it's definitely worth experimenting

2) Given that `SortOrderValues` is a constant, should we have a convention to just name it `SORT_ORDER` (and call `SORT_ORDER.ascending`?). This is also a bit shorter than adding the `Values` suffix everywhere. :)

**Comment by domoritz** [comment]
That's what I like about the approach in Vega-Lite. The type and the enum (namespace) have the same name.

**Comment by kanitw** [comment]
Oh right, I wonder if the same approach works with the transformer.

**Comment by domoritz** [comment]
Should we close this issue for now?
